src/concrete/ml/quantization/base_quantized_op.py

```diff
@@ -933,39 +933,62 @@ def make_output_quant_parameters(
         )
 
     def cnp_round(
-        self, x: Union[numpy.ndarray, fhe.tracing.Tracer], calibrate_rounding: bool
+        self,
+        x: Union[numpy.ndarray, fhe.tracing.Tracer],
+        calibrate_rounding: bool,
+        key: Optional[str] = None,
     ) -> numpy.ndarray:
         """Round the input array to the specified number of bits.
 
         Args:
             x (Union[numpy.ndarray, fhe.tracing.Tracer]): The input array to be rounded.
             calibrate_rounding (bool): Whether to calibrate the rounding
                 (compute the lsbs_to_remove)
+            key (Optional[str]): The key to access the appropriate
+                lsbs_to_remove value in the dictionary.
 
         Returns:
             numpy.ndarray: The rounded array.
         """
 
-        # Rounding is applied only if specified by user
-        if self.rounding_threshold_bits is not None:
-            if calibrate_rounding:
-                assert_true(
-                    not isinstance(x, fhe.tracing.Tracer),
-                    "Can't compute lsbs_to_remove at compilation time.",
-                )
-                assert_true(
-                    self.lsbs_to_remove is None,
-                    "Rounding has already been calibrated.",
-                )
+        # Initialize lsbs_to_remove if not already set
+        if not hasattr(self, "lsbs_to_remove") or self.lsbs_to_remove is None:
+            self.lsbs_to_remove = {} if key else 0
 
-                current_n_bits_accumulator = compute_bits_precision(x)
-                self.lsbs_to_remove = current_n_bits_accumulator - self.rounding_threshold_bits
+        if self.rounding_threshold_bits is not None and calibrate_rounding:
+            # Compute new_value only when calibration is True
+            current_n_bits_accumulator = compute_bits_precision(x)
+            new_value = current_n_bits_accumulator - self.rounding_threshold_bits
 
-            # mypy
-            assert self.lsbs_to_remove is not None
+            assert_true(
+                not isinstance(x, fhe.tracing.Tracer),
+                "Can't compute lsbs_to_remove at compilation time.",
+            )
+
+            # Key-based logic (dictionary)
+            if key:
```

Comment:
if we keep this dict / int duality, then better change this to `if key is not None` !

Reply:
It's the same isn't it?